build_tools/docker/manage_images.py

```diff
@@ -0,0 +1,249 @@
+    print(inspect_process.stderr, end='')
+    sys.exit(inspect_process.returncode)
+  _, repo_digest = inspect_process.stdout.strip().split('@')
+  return repo_digest
+
+
+def update_rbe_reference(digest, dry_run=False):
+  print('Updating WORKSPACE file for rbe-toolchain')
+  for line in fileinput.input(files=['WORKSPACE'], inplace=(not dry_run)):
+    if line.strip().startswith('digest ='):
+      print(re.sub('sha256:[a-zA-Z0-9]+', digest, line), end='')
+    else:
+      print(line, end='')
+
+
+def update_references(image_name, digest, dry_run=False):
+  print(f'Updating references to {image_name}')
+
+  # Just hardcode this oddity
+  if image_name.rsplit('/', maxsplit=1)[-1] == 'rbe-toolchain':
+    update_rbe_reference(digest, dry_run=dry_run)
+
+  grep_command = ['git', 'grep', '-l', f'{image_name}@sha256']
+  grep_process = subprocess.run(grep_command,
+                                capture_output=True,
+                                timeout=5,
+                                text=True)
+  if grep_process.returncode > 1:
+    print(f'{" ".join(grep_command)} '
+          f'failed with exit code {grep_process.returncode}')
+    sys.exit(grep_process.returncode)
+  if grep_process.returncode == 1:
+    print(f'Found no references to {image_name}')
+    return
+
+  files = grep_process.stdout.split()
+  print(f'Updating references in {len(files)} files: {files}')
+  for line in fileinput.input(files=files, inplace=(not dry_run)):
+    print(re.sub(f'{image_name}@sha256:[a-zA-Z0-9]+', f'{image_name}@{digest}', line),
+          end='')
+
+
+if __name__ == '__main__':
+  args = parse_arguments()
+
+  # Ensure the user has the correct authorization if they try to push to GCR.
+  if args.push:
+    if stream_command(['which', 'gcloud']) != 0:
+      print('gcloud not found.'
+            ' See https://cloud.google.com/sdk/install for installation.')
+      sys.exit(1)
+    check_stream_command(['gcloud', 'auth', 'configure-docker'],
+                         dry_run=args.dry_run)
+
+
+  images_to_process = get_ordered_images_to_process(args.images)
+  print(f'Also processing dependent images. Will process: {images_to_process}')
+
+  for image in images_to_process:
+    print(f'Processing image {image}')
+    image_name = os.path.join(IREE_GCR_URL, image)
+    image_tag = f'{image_name}:{args.tag}'
+    image_path = os.path.join(DOCKER_DIR, image.replace('-', '_'))
+
+    if args.pull:
+      check_stream_command(['docker', 'pull', image_tag], dry_run=args.dry_run)
+
+    if args.build:
+      check_stream_command(['docker', 'build', '--tag', image_tag, image_path],
+                           dry_run=args.dry_run)
+
+    if args.push:
+      check_stream_command(['docker', 'push', image_tag], dry_run=args.dry_run)
+
+    if args.update_references:
+      digest = get_repo_digest(image_tag)
+      # Just hardcode this oddity
+      if image == 'rbe-toolchain':
+        update_rbe_reference(digest, dry_run=args.dry_run)
```

Comment:
This looks like a simpler version of `[189, 191)` – did they accidently escape deletion?

Reply:
They did. Deleted.